### PR TITLE
Scipio controllers: deadline queue

### PR DESCRIPTION
**Pull request by glommer**
### What does this PR do?

This PR adds Scipio's first controller. A controller is an entity that can dynamically and automatically adjust shares of a task queue lower or higher so that the tasks in that task queue proceed accordingly to some user-defined criteria.

The first of such controllers is the DeadlineQueue: It is a version of a work queue, in which each element is given a deadline. The goal of the controller is to minimize the impact of running the work queue by reducing shares as much as possible so long as each element finishes within its deadline

It comes with an interactive example you can play with!

### Motivation

Process controllers were used extensively in Scylla to control background processes such as buffer flushes and LSM compactions, making sure that the database had stable latencies throughout its operation. However, those were done
at the application level (Scylla), as opposed to the framework level (seastar)

The challenge I set myself to solve for Scipio was to do this in a generic enough fashion so that it can live in the library and
be easily consumable by application developers.

This means a considerable amount of effort will be poured into wrapping up the controllers in a series of APIs that makes sense for general purpose use.

The first such API is the DeadlineQueue: a workqueue-like structure that queues items implementing a particular trait. (DeadlineSource)

Items in that queue tell the controller how much work they are supposed to do (in user-defined units) and how how much work they have done already. Items execute sequentially. The controller does the rest.

### CODE REVIEW

**Comment by glommer** [comment]
Heads up: coming up with unit tests for controllers is *HARD* but quitters never win, and winners never quit.
I'll do my best to test at least some boundary conditions.

In the meantime the example should be super nice to play with and is what I have been mostly using for test purposes.

File: examples/deadline_writer.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    deadline.push_work(test).await.unwrap()
+                }}).detach();
+
+                Local::later().await;
+
+                let stop = Rc::new(Cell::new(false));
+                let hog = competing_cpu_hog(stop.clone(), cpuhog_tq);
+                let dur = writer.await.unwrap();
+                println!("Finished writing in {}", Colour::Green.paint(format!("{:#.2?}", dur)));
+                stop.replace(true);
```
Comment (matklad) [inline examples/deadline_writer.rs:201]: Is there any particular reason to use `.replace` over `.set`? This feels slightly unusual.

File: scipio/src/controllers/deadline_queue.rs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    /// What type is returned by the [`action`] method
+    ///
+    /// [`action`]: trait.DeadlineSource.html#method.action
+    type Output;
+
+    /// Returns a [`Duration`] indicating when we would like this operation to complete.
+    ///
+    /// It is calculated from the point of Queueing, not from the point in which the operation
+    /// starts.
+    fn deadline(&self) -> Duration;
```
Comment (matklad) [inline scipio/src/controllers/deadline_queue.rs:36]: Without reading the code closely, the signature feels weird. I'd expect `fn deadline(&self) -> Instant`

File: scipio/src/controllers/deadline_queue.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    ///
+    /// For example, you can buffer all updates and just inform that processed_units == total_units
+    /// at the end of the process, but then the controller would be a step function.
+    ///
+    /// [`total_units`]: trait.DeadlineSource.html#method.total_units
+    fn processed_units(&self) -> u64;
+}
+
+impl<T> Debug for dyn DeadlineSource<Output = T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```
Comment (matklad) [inline scipio/src/controllers/deadline_queue.rs:77]: Minor style nit, but I find the following way to write this to be aesthetically nicest:
```suggestion
impl<T> fmt::Debug for dyn DeadlineSource<Output = T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

File: scipio/src/controllers/deadline_queue.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+            adjustment_period,
+            last_error: Cell::new(0.0),
+            min_shares: Cell::new(1),
+        }
+    }
+
+    fn admit(&self, source: Rc<dyn DeadlineSource<Output = T>>) -> io::Result<()> {
+        let expiration = Instant::now()
+            .checked_add(source.deadline())
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, ""))?;
```
Comment (matklad) [inline scipio/src/controllers/deadline_queue.rs:255]:
```suggestion
            .ok_or_else(|| io::Error::from(io::ErrorKind::InvalidInput))?;
```

File: scipio/src/controllers/deadline_queue.rs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+            if adjustment_period < Duration::from_millis(100) {
+                Latency::Matters(adjustment_period)
+            } else {
+                Latency::NotImportant
+            }
+        };
+
+        let tq = Local::create_task_queue(Shares::Dynamic(queue.clone()), lat, name);
+        let (sender, receiver): (LocalSender<QueueItem<T>>, LocalReceiver<QueueItem<T>>) =
+            LocalChannel::new_bounded(1);
```
Comment (matklad) [inline scipio/src/controllers/deadline_queue.rs:369]: channels APIs usually make the ctor function top-level. That way, you'll have a nice place to put a type parameter once: `local_channel::new_bounded::<QueueItem<T>>(1)`

File: examples/deadline_writer.rs (hunk 1 above)
Comment (glommer) [reply examples/deadline_writer.rs:201]: no, just force of habit as sometimes I do want the value back then end up always using replace from muscle memory. Will change.

File: scipio/src/controllers/deadline_queue.rs (hunk 2 above)
Comment (glommer) [reply scipio/src/controllers/deadline_queue.rs:36]: right, so maybe the name needs to be changed.
This represents how long the operation is supposed to run for - we use that to calculate speed later so this has to be a duration.

When we enqueue it we add that to now() to obtain an instant, which is then placed in the queue collection.

Maybe just call it `duration` ?

File: scipio/src/controllers/deadline_queue.rs (hunk 3 above)
Comment (glommer) [reply scipio/src/controllers/deadline_queue.rs:77]: agree

File: scipio/src/controllers/deadline_queue.rs (hunk 5 above)
Comment (glommer) [reply scipio/src/controllers/deadline_queue.rs:369]: so just a free function without an associated struct?
I'll change that in the channel PR.

File: scipio/src/channels/local_channel.rs [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+/// [`next`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.next
+/// [`StreamExt`]: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html
+pub struct LocalChannel<T> {
+    state: Rc<RefCell<State<T>>>,
+}
+
+impl<T> LocalChannel<T> {
+    fn new(capacity: ChannelCapacity) -> (LocalSender<T>, LocalReceiver<T>) {
+        let channel = match capacity {
+            ChannelCapacity::Unbounded => VecDeque::new(),
```
Comment (duarten) [inline scipio/src/channels/local_channel.rs:169]: Is there a good use case for unbounded channels?

File: scipio/src/channels/local_channel.rs [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+        }
+
+        state.channel.push_back(item);
+        if !state.recv_waiters.is_empty() {
+            let waiters = std::mem::replace(&mut state.recv_waiters, Vec::new());
+            drop(state);
+
+            for w in waiters {
+                w.wake();
+            }
```
Comment (duarten) [inline scipio/src/channels/local_channel.rs:213]: Shouldn't it wake just one?

File: scipio/src/channels/local_channel.rs [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+        }
+    }
+}
```
Comment (duarten) [inline scipio/src/channels/local_channel.rs:394]: Can a `LocalSender` be dropped while there are still `send_waiters`? Something like calling `send_eventually` without awaiting the result.

File: scipio/src/controllers/deadline_queue.rs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+impl<T> PriorityBump<T> {
+    fn new(queue: Rc<InnerQueue<T>>) -> PriorityBump<T> {
+        queue.min_shares.replace(250);
+        PriorityBump { queue }
+    }
+}
+
+impl<T> Drop for PriorityBump<T> {
+    fn drop(&mut self) {
+        self.queue.min_shares.replace(1);
```
Comment (duarten) [inline scipio/src/controllers/deadline_queue.rs:111]: I guess no one will do 

```ts
let pb = PriorityBump::new(q)
let pb2 = PriorityBump::new(pb.queue)
drop(pb) // back to 1
```

File: scipio/src/controllers/deadline_queue.rs [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+        self.accumulated_error.replace(acc);
+        let delta_error = error - self.last_error.get();
+        self.last_error.replace(error);
+
+        // How did we pick our constants:
+        //  * As with any stable PI controller we want the bulk of our gain to come from P.
+        //  * As we normalize the maximum error to 1 we know that the gain should be at most 1000
+        //  * physically, Ki can be expressed as Kp / Tau where Tau is a time constant, roughly
+        //    equivalent to how many periods need to pass for the integral term to generate the
+        //    same gain as the proportional term. We set that too 6 so the controller is not too
```
Comment (duarten) [inline scipio/src/controllers/deadline_queue.rs:214]: s/too/to

File: scipio/src/controllers/deadline_queue.rs [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+    //    be comparable as we accumulate artifacts of the delta_t calculation (as we'll
+    //    never in practice keep delta_t constant)
+    //
+    //  The way we'll solve this is by expressing an alternate error E(T) which is
+    //  essentially the integral of e(t) in time:
+    //
+    //    E(t) = 1 - units_processed / units_expected.
+    //
+    //  That is easy to compute, as it is essentially the total count of units for
+    //  all items in the queue, both processed an expected.
```
Comment (duarten) [inline scipio/src/controllers/deadline_queue.rs:167]: s/an/and

File: scipio/src/controllers/deadline_queue.rs [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+        self.results.pop_front().await?
+    }
+}
+
+#[derive(Debug)]
+/// Scipio's scheduler is based on [`Shares`]: the more shares, the more resources the task
+/// will receive.
+///
+/// There are situations however in which we don't want shares to grow too high: for instance,
+/// a background process that is flushing a file to storage. If it were to run a full speed,
```
Comment (duarten) [inline scipio/src/controllers/deadline_queue.rs:285]: s/a/at

File: scipio/src/controllers/deadline_queue.rs [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+///   between the application of the control decision and its effects being seen. In our case, the
+///   scheduler may not schedule us for a long time, especially if the shares are low.
+/// * Control loops work better the faster and smoother the response is. Let's use an example flushing a
+///   file: you may be moving data to the file internal buffers but they are not *flushed* to the
+///   media. When the data is finally flushed a giant bubble is inserted into the control loop. The
+///   characteristics of the system will radically change. Contract that for instance with O_DIRECT
+///   files, where writing to the file means writing to the media: smooth and fast feedback!
+///
+///   The moral of the story is:
+///    * do not use this with buffered files or other buffered sinks where where the real physical response
```
Comment (duarten) [inline scipio/src/controllers/deadline_queue.rs:316]: s/where/

**Review by duarten — COMMENTED** [review]
very cool :)

File: scipio/src/channels/local_channel.rs (hunk 6 above)
Comment (matklad) [reply scipio/src/channels/local_channel.rs:169]: Non-DAG communication graphs (actor systems) need at least some channels to be unbounded, otherwise they deadlock: https://trio.discourse.group/t/sizing-the-channel-deadlock-freedom-vs-back-pressure/311

Comment (glommer) [reply scipio/src/channels/local_channel.rs:169]: More importantly, this is not an opinion that I would like Scipio to have.
Other crates for channels have this which indicates that some people have uses for unbounded channels.


Scipio should strive for being generic and making many use cases possible.

File: scipio/src/channels/local_channel.rs (hunk 7 above)
Comment (glommer) [reply scipio/src/channels/local_channel.rs:213]: That's a very good question. It's clearly not wrong to wake them all since they'll just be put to sleep.
Intuitively it feels to me that if we are pushing and consuming a single item, a single notification would be enough and you are right.

I need to convince myself that a lost notification is not possible in the parallel executors, though.

File: scipio/src/controllers/deadline_queue.rs (hunk 9 above)
Comment (glommer) [reply scipio/src/controllers/deadline_queue.rs:111]: I would consider this a bug.
Notice how you can't even specify the priority you want in the API, so this is not like a verbose flag that you can pass many.

File: scipio/src/channels/local_channel.rs (hunk 8 above)
Comment (glommer) [reply scipio/src/channels/local_channel.rs:394]: I spent some time thinking about this.

I don't think it should be possible:
- Rust futures are lazy, so there is no such thing as "not waiting on them". If you don't .await they don't execute.
- For that situation to trigger you have to spawn a new detached task and have the `LocalSender` and the send operation running in different tasks
- But for you to pass that into a spawn-like method you need an `Rc` pointer: spawn-like methods require static lifetime for all references passed to it.
- Which means if you drop at this point you'd be dropping the reference count. The senders would keep the refcount up and prevent an early drop.

File: scipio/src/channels/local_channel.rs (hunk 7 above)
Comment (duarten) [reply scipio/src/channels/local_channel.rs:213]: Not totally clear on what the parallel executors are, but in general, a _check - signal sleep - redo check - sleep_ pattern works well when there's parallelism.

File: scipio/src/channels/local_channel.rs (hunk 6 above)
Comment (duarten) [reply scipio/src/channels/local_channel.rs:169]: It could have that opinion, similar to how Go has it. Unbounded queues are usually anathema for latency-sensitive solutions (unbounded queues do solve deadlocks in CSP systems, in a similar way that a Big Lock can solve data races).

Comment (glommer) [reply scipio/src/channels/local_channel.rs:169]: you are right about the latency issue but I would like scipio to be useful to build latency sensitive applications, but *not only* to build latency sensitive applications.

Sometimes the complexity of the application is elsewhere.

File: scipio/src/channels/local_channel.rs (hunk 7 above)
Comment (glommer) [reply scipio/src/channels/local_channel.rs:213]: Actually I think we should be waking just one here. Otherwise imagine a blocked channel with 1,000,000 waiters that gets 1,000,000 elements pushed: it will call `wake` a lot as it becomes quadratic.

File: scipio/src/controllers/deadline_queue.rs (hunk 2 above)
Comment (glommer) [reply scipio/src/controllers/deadline_queue.rs:36]: Called it `expected_duration`

**Comment by glommer** [comment]
rebased. All the dependencies are now in, so this is just the controller code.

**Comment by glommer** [comment]
forced pushed a version:
* added unit tests, especially for the boundary conditions
* improved the example
* changed the API so this is easily usable behind `Rc` through interior mutability (no `&mut self`) for `push`.

File: scipio/src/controllers/deadline_queue.rs [hunk 14]
```diff
@@ -0,0 +1,637 @@
...
+        }
+    }
+
+    fn admit(&self, source: Rc<dyn DeadlineSource<Output = T>>) -> io::Result<()> {
+        let expiration = Instant::now()
+            .checked_add(source.expected_duration())
+            .ok_or_else(|| io::Error::from(io::ErrorKind::InvalidInput))?;
+        expiration
+            .checked_duration_since(self.last_admitted.get())
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, ""))?;
```
Comment (matklad) [inline scipio/src/controllers/deadline_queue.rs:256]:
```suggestion
            .ok_or_else(|| io::Error::from(io::ErrorKind::InvalidInput))?;
```
